cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mesh
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+
+	"istio.io/operator/pkg/manifest"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	meshInfoVersion "istio.io/pkg/version"
+
+	goversion "github.com/hashicorp/go-version"
+
+	"istio.io/operator/pkg/compare"
+	pkgversion "istio.io/operator/pkg/version"
+	opversion "istio.io/operator/version"
+)
+
+const (
+	// The maximum duration the command will wait until the apply deployment reaches a ready state
+	upgradeWaitSecWhenApply = 300
+	// The time that the command will wait between each check of the upgraded version.
+	upgradeWaitSecCheckVerPerLoop = 10
+	// The maximum number of attempts that the command will check for the upgraded version.
+	upgradeWaitCheckVerMaxAttempts = 60
+)
+
+var (
+	l *logger
```

Comment:
Rather than using l as a global var in the code, it's better to define some helper functions that wrap the log function and do a lazy init of the logger. 
Alternative (worse imo) would be to call this something more global sounding (l would be used for naming local vars) and initialize it in an init() function.

Reply:
I can not get the specific work items here. I will follow the current patterns in the repo, making logger as a parameter when required in a function.

We can refactor the log into a repo, and do lazy initialization there.